storage/aptosdb/src/pruner/state_store/mod.rs

```diff
@@ -121,96 +125,176 @@ where
-                    existing_schema_batch.delete::<S>(&index)
-                })?;
+    ) -> Result<Version> {
+        let mut min_readable_version = min_readable_version;
+        let mut target_version_for_this_batch = min_readable_version;
+        while target_version_for_this_batch <= target_version {
+            self.record_progress(target_version_for_this_batch);
+            let next_version =
+                self.prune_top_levels(min_readable_version, target_version_for_this_batch)?;
+            self.prune_shards(target_version_for_this_batch, batch_size)?;
+            min_readable_version = target_version_for_this_batch;
+            if let Some(next_version) = next_version {
+                target_version_for_this_batch = next_version;
             } else {
-                let batch = SchemaBatch::new();
-                indices.into_iter().try_for_each(|index| {
-                    batch.delete::<JellyfishMerkleNodeSchema>(&index.node_key)?;
-                    batch.delete::<S>(&index)
-                })?;
+                break;
+            }
+        }
+        Ok(min_readable_version)
+    }
 
-                self.save_min_readable_version(new_min_readable_version, &batch)?;
+    fn prune_top_levels(
+        &self,
+        min_readable_version: Version,
+        target_version: Version,
+    ) -> Result<Option<Version>> {
+        let batch = SchemaBatch::new();
+        let next_version = self.prune_state_merkle_shard(
+            self.state_merkle_db.metadata_db(),
+            min_readable_version,
+            target_version,
+            usize::max_value(),
+            &batch,
+        )?;
+        batch.put::<DbMetadataSchema>(&S::tag(None), &DbMetadataValue::Version(target_version))?;
+        self.state_merkle_db.metadata_db().write_schemas(batch)?;
 
-                // TODO(grao): Support sharding here.
-                self.state_merkle_db.metadata_db().write_schemas(batch)?;
-            }
+        Ok(next_version)
+    }
 
-            // TODO(zcc): recording progress after writing schemas might provide wrong answers to
-            // API calls when they query min_readable_version while the write_schemas are still in
-            // progress.
-            self.record_progress_impl(new_min_readable_version, is_end_of_target_version);
-            Ok(new_min_readable_version)
+    fn prune_single_shard(
+        &self,
+        shard_id: u8,
+        target_version: Version,
+        batch_size: usize,
+    ) -> Result<()> {
+        let _timer = OTHER_TIMERS_SECONDS
+            .with_label_values(&["state_merkle_pruner___prune_single_shard"])
+            .start_timer();
+        let shard_min_readable_version = self.get_shard_progress(shard_id);
+        if shard_min_readable_version != target_version {
+            assert_lt!(shard_min_readable_version, target_version);
+            self.update_shard_progress(shard_id, target_version);
```

Comment:
Ideally, do we want to update shard progress once the prune is done?

Reply:
The db progress is updated after, the in memory one is updated before, to tell the upper layer the data is not readable.

Reply:
What I said above has changed now. Now we manage the "min_readable_version" in the pruner manager (updated before the work, and this is the value we serve to outside), and we manager the "progress" in the pruner structs (updated after the work is done).